parsl/app/errors.py

```diff
@@ -128,9 +136,12 @@ def reraise(self):
         reraise(t, v, tb)
 
 
-def wrap_error(func):
+# TODO: [typing] I don't think this is correct. We need to constrain the type of the
+# wrapper to that of the wrapped function, whereas this specification makes the wrapper
```

Comment:
I added a stub in mypy-stubs for @typeguard.typechecked annotations a while ago, because I wanted a typeguard decorated function to have the same type signature as the original pre-decorated function.

That code lives in mypy-stubs/typeguard.pyi

That feels to me similar-ish to what wrap_error is trying to do (but not in decorator form).

It looks like this:
```
from typing import TypeVar

Func = TypeVar('Func')

def typechecked(f: Func) -> Func: ...
```

Reply:
I had a bit of a play with this. One problem is that the wrapped function can return *more* than the original function - it might return a RemoteExceptionWrapper, rather than a value of the original returned type (so if the original return type was T, the new return type is Union[T, RemoteExceptionWrapper].

Reply:
I was thinking that the typevar trick won't work since the wrapper returned by the decorator should have a proper signature. Bu then it looks like it's possible to simply not typecheck that part:

```python
T = TypeVar('T')

def decorator(f: T) -> T:
        def w(*args, **kwargs):
                return f(*args, **kwargs)
        return w  # type: ignore
```

So I guess that and Union[T, ...] should work here.

Reply:
i went down a rathole reading about this and playing yesterday and the short answer seems to be no, the longer answer being "no, but the typed python community is slowly working towards it; you can write a hacky mypy plugin in the meantime".

This seems to be the big relevant issue on mypy:
 
https://github.com/python/mypy/issues/3157

so if you don't figure out a pleasing solution, I guess add that issue into the TODO comment.

I was a bit concerned in my testing yesterday about how hard it was to get a type error to occur when I had got the types wrong in particular ways, so grr.